gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -348,13 +480,28 @@ public void onNext(Object response) {
       public void onError(Throwable t) {
         Status s = Status.fromThrowable(t);
         String statusDesc = s == null ? "" : s.getDescription();
-        error =
-            new IOException(
-                String.format(
-                    "Caught exception for '%s', while uploading to uploadId %s at writeOffset %d."
-                        + " Status: %s",
-                    resourceId, uploadId, writeOffset, statusDesc),
-                t);
+
+        if (t.getClass() == StatusException.class || t.getClass() == StatusRuntimeException.class) {
+          Code code =
+              t.getClass() == StatusException.class
+                  ? ((StatusException) t).getStatus().getCode()
+                  : ((StatusRuntimeException) t).getStatus().getCode();
+          if (code == Code.DEADLINE_EXCEEDED
+              || code == Code.RESOURCE_EXHAUSTED
+              || code == Code.INTERNAL
+              || code == Code.UNAVAILABLE) {
+            transientError = t;
+          }
+        }
+        if (transientError == null) {
+          nonTransientError =
+              new Throwable(
```

Comment:
Set `t` as a cause of this `Throwable` and change exception type of all thrown exceptions to `IOException` - I think this is more appropriate.

Reply:
Done.